returnn/torch/updater.py

```diff
@@ -223,13 +225,35 @@ def step(self, *, grad_scaler: Optional[torch.cuda.amp.GradScaler] = None):
             gradient_noise_(self.network.parameters(), self._grad_noise)
         if self._grad_clip:
             torch.nn.utils.clip_grad_value_(self.network.parameters(), self._grad_clip)
-        if self._grad_clip_global_norm:
-            torch.nn.utils.clip_grad_norm_(self.network.parameters(), self._grad_clip_global_norm)
+
+        has_invalid_gradient = False
+        if self._num_allowed_invalid_gradient_steps is not None:
+            if self._grad_clip_global_norm:
+                norm = torch.nn.utils.clip_grad_norm_(self.network.parameters(), self._grad_clip_global_norm)
+            else:
+                norm = torch.nn.utils.get_total_norm(self.network.parameters())
+            has_invalid_gradient = torch.isnan(norm) or torch.isinf(norm)
+            if has_invalid_gradient:
+                self._num_invalid_gradients += 1
+                if self._num_invalid_gradients >= self._num_allowed_invalid_gradient_steps:
+                    raise RuntimeError(
+                        f"Got {self._num_invalid_gradients} invalid gradients in succession, abort training"
+                    )
+                else:
+                    invalid_grads_left = self._num_allowed_invalid_gradient_steps - self._num_invalid_gradients
+                    print(
+                        f"Invalid gradient in step {self._current_train_step}, skipping. {invalid_grads_left:02d} "
+                        f"subsequent broken steps left until training is aborted."
+                    )
+            else:
+                self._num_invalid_gradients = 0
 
         if grad_scaler is not None:
-            grad_scaler.step(self.optimizer)
+            if not has_invalid_gradient:
+                grad_scaler.step(self.optimizer)
+            # updater needs to be called even if we discard the update due to an invalid gradient
```

Comment:
```suggestion
            # update needs to be called even if we discard the update due to an invalid gradient
```
Why? Any reference? Is this always correct?